File: frontend/src/services/busca.service.ts
import { apiRequest } from "./api";
import type { BuscaLicitacaoFilters, BuscaLicitacoesResponseType } from "../types/licitacao.types";

export async function buscarLicitacoes(
  filters: BuscaLicitacaoFilters,
): Promise<BuscaLicitacoesResponseType> {
  return apiRequest<BuscaLicitacoesResponseType>("/busca/licitacoes", {
    query: {
      q: filters.buscar_por,
      buscar_por: filters.buscar_por,
      portais: filters.portais.join(","),
      numero_oportunidade: filters.numero_oportunidade,
      objeto_licitacao: filters.objeto_licitacao,
      orgao: filters.orgao,
      empresa: filters.empresa,
      sub_status: filters.sub_status,
      tipo_instrumento_convocatorio: filters.tipo_instrumento_convocatorio,
      unidade: filters.unidade,
      estado: filters.estado,
      municipio: filters.municipio,
      esfera: filters.esfera,
      poder: filters.poder,
      fonte_orcamentaria: filters.fonte_orcamentaria,
      margem_preferencia: filters.margem_preferencia,
      conteudo_nacional: filters.conteudo_nacional,
      modalidade: filters.modalidade,
      tipo_fornecimento: filters.tipo_fornecimento.join(","),
      familia_fornecimento: filters.familia_fornecimento.join(","),
      data_inicio: filters.data_inicio,
      data_fim: filters.data_fim,
      pagina: filters.pagina ?? 1,
      page_size: 10,
    },
  });
}

export async function buscarLicitacoesInteligente(payload: {
  objetivo: string;
  portais: string[];
  estado?: string;
  municipio?: string;
  pagina?: number;
}): Promise<BuscaLicitacoesResponseType> {
  return apiRequest<BuscaLicitacoesResponseType>("/busca/licitacoes/inteligente", {
    method: "POST",
    body: JSON.stringify({
      objetivo: payload.objetivo,
      portais: payload.portais,
      estado: payload.estado || null,
      municipio: payload.municipio || null,
      pagina: payload.pagina ?? 1,
      page_size: 10,
    }),
  });
}

File: backend/app/services/busca_service.py
from sqlalchemy.orm import Session

from app.schemas.busca import BuscaLicitacoesResponse
from app.services.busca import BuscaAggregator, SearchQuery


class BuscaService:
    def __init__(self, db: Session) -> None:
        self.db = db
        self.aggregator = BuscaAggregator(db)

    async def buscar_licitacoes(
        self,
        q: str | None,
        buscar_por: str | None,
        portais: list[str],
        numero_oportunidade: str | None,
        objeto_licitacao: str | None,
        orgao: str | None,
        empresa: str | None,
        sub_status: str | None,
        tipo_instrumento_convocatorio: str | None,
        unidade: str | None,
        estado: str | None,
        municipio: str | None,
        esfera: str | None,
        poder: str | None,
        fonte_orcamentaria: str | None,
        margem_preferencia: str | None,
        conteudo_nacional: str | None,
        modalidade: str | None,
        tipo_fornecimento: list[str],
        familia_fornecimento: list[str],
        data_inicio: str | None,
        data_fim: str | None,
        pagina: int,
        page_size: int = 10,
    ) -> BuscaLicitacoesResponse:
        query = SearchQuery(
            q=q,
            buscar_por=buscar_por,
            portais=portais,
            numero_oportunidade=numero_oportunidade,
            objeto_licitacao=objeto_licitacao,
            orgao=orgao,
            empresa=empresa,
            sub_status=sub_status,
            tipo_instrumento_convocatorio=tipo_instrumento_convocatorio,
            unidade=unidade,
            estado=estado,
            municipio=municipio,
            esfera=esfera,
            poder=poder,
            fonte_orcamentaria=fonte_orcamentaria,
            margem_preferencia=margem_preferencia,
            conteudo_nacional=conteudo_nacional,
            modalidade=modalidade,
            tipo_fornecimento=tipo_fornecimento,
            familia_fornecimento=familia_fornecimento,
            data_inicio=data_inicio,
            data_fim=data_fim,
            pagina=pagina,
            page_size=page_size,
        )
        return await self.aggregator.search(query)
